### ISSUE TITLE
Invalid date/time values generated (DST)

### ISSUE THREAD

**Issue by telkins** [fzaninotto/Faker#1078]
From time to time we get errors when attempting to insert Faker-generated data into our MySQL database.  We're specifically using `dateTimeThisYear()`.

Here's the error: `SQLSTATE[22007]: Invalid datetime format: 1292 Incorrect datetime value: '2016-03-27 02:08:58'`

As you can see, at first glance, the date/time looks valid, but it's actually between 2 and 3 AM on the early Sunday morning where time leaps forward from 2 to 3 AM.  So, any time between 2 and 3 AM doesn't actually exist.

The column in question is a typical MySQL `datetime` field.

We're running Faker 1.6.0, PHP 7.0.8, Laravel 5.3.19, and MySQL 5.7.13-0ubuntu0.16.04.2, not that it really matters.

I'm not 100% certain that this is a Faker problem because, as I understand it, there are no time changes in UTC, the MySQL server in question is set to use the system timezone, and the system timezone appears to be CEST.  Although, as I mentioned, we're using `dateTimeThisYear()`, so there's no way (as I understand it) to specify the timezone.  Whatever the case, if someone knows better, please let me know.

**Comment by curry684** [comment]
This is not Faker's issue as I see it. Faker is just generating random data within a rigid set of rules, so in the case of a DateTime that means it follows rules like `0 <= hour < 24`. It's completely out of scope for Faker to consider the validity of that time within any more specific situation than Coordinated Time (UTC), which it automatically is according to those constraints. With good reason, as timezone specific handling is a [completely ridiculous field of expertise in IT](http://stackoverflow.com/a/6841479/1729885) with tens of thousands of exceptions in many timezones. Catering to all of them isn't the job of a simple PHP faker library, especially given that there are [wonderful standard libraries](https://packages.debian.org/stretch/tzdata) that contain 2.8MB of data describing [every known exception ever in every timezone](https://packages.debian.org/stretch/all/tzdata/filelist).

So your problem is not in Faker, but in configuring your MySQL not to use UTC. Which is actually not that bad a choice by definition, assuming that MySQL is only ever going to get time data prevalidated to those constraints you imposed on it willingly. That's where this all starts to crash - the point where you the consumer of a generic fake data generator are blindly submitting unvalidated data into your database. You would have much more frequent problems if you didn't happen to live in CE(S)T which is one of the most stable timezones over the past century.

In the end the solution is pretty simple - accept that Faker gives you UTC and convert it to the right zone:
```
$dateTime = new \DateTime($generatedDateTime, new \DateTimeZone('UTC'));
$localizedDateTime = $dateTime->setTimeZone(new \DateTimeZone('Europe/Amsterdam'));
```
Or use [`date_default_timezone_get`](http://php.net/manual/en/function.date-default-timezone-get.php) in the second line to get PHP's configured timezone, assuming that's correctly configured.

**Comment by telkins** [comment]
@curry684 Thanks for the response.  It was informative and it provided a potential solution that I hadn't previously considered.  I think I'll implement it.

On a related note, I think it might be worth suggesting a feature/improvement.  Since several of the date/time related methods offer a `timezone` parameter, why not introduce a similar parameter to the other date/time related methods, where it makes sense.  In this case, it would be beneficial to introduce a `timezone` parameter to `dateTimeThisYear()`.  It could provide the solution you suggested, but as a part of the Faker library.  It could continue to produce UTC date/time values by default, but allow a user like myself to ask Faker to perform the conversion you suggested, like it does with several of the other methods.

Thoughts?

**Comment by telkins** [comment]
@curry684 Just a quick follow-up to my last post....I took a quick look at the source code and it appears that `dateTimeThisYear()` calls `dateTimeBetween()`, which takes a `timezone` parameter as its third parameter.  It seems like it would be a fairly simple change to add `timezone` to `dateTimeThisYear()` (and others like it) and pass it along to `dateTimeBetween()`.

That's it.  I'd still like to hear your thoughts on all this.  :-)

**Comment by curry684** [comment]
I think that makes sense if current defaults are retained. Actually many locale definitions could be extended to override the default, for example nl_NL could happily default to CEST rules.

**Comment by telkins** [comment]
@curry684 Thanks for the feedback.  I'm relatively new to Faker.  What do you suggest?  Submit an issue requesting the feature and/or a PR with the desired changes?  Or something else...?

I'm happy to contribute, but I'm also happy to let others who are more familiar with the library do what needs to be done.  :-)

**Comment by curry684** [comment]
My "Faker contributor" badge is also from a single PR in the Dutch locales ;)

@fzaninotto what is your take on this?

**Comment by telkins** [comment]
@curry684 I've made some changes, added some tests, and created a PR: https://github.com/fzaninotto/Faker/pull/1090

Thoughts...?

**Comment by fzaninotto** [comment]
Fixed by #1090
